File: examples/demo3.py
'''
Lower-level interface demo that illustrates creating and running gadget chains
of 3 types:
- acquisition processing chain
- reconstruction chain
- image processing chain
'''

import argparse
import os
import sys

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

parser = argparse.ArgumentParser(description = \
'''
Lower-level interface demo that illustrates creating and running gadget chains
of 3 types:
(i) acquisition processing chain
(ii) reconstruction chain
(iii) image processing chain
''')
parser.add_argument\
('filename', nargs='?', default = 'testdata.h5', \
 help = 'raw data file name (default: testdata.h5)')
args = parser.parse_args()                                 

def main():

    # acquisitions will be read from an HDF file args.filename
    input_data = MR_Acquisitions(args.filename)

    # define gadgets
    gadget1 = Gadget('RemoveROOversamplingGadget')
    gadget2 = Gadget\
        ('SimpleReconGadgetSet(trigger_dimension=repetition,split_slices=true)')
    gadget3 = Gadget('ExtractGadget')

    # build acquisition pre-processing chain
    acq_proc = AcquisitionsProcessor()
    acq_proc.add_gadget('g1', gadget1)
    print('processing acquisitions...')
    preprocessed_data = acq_proc.process(input_data)

    # create reconstruction object
    recon = ImagesReconstructor()
    recon.add_gadget('g2', gadget2)
    # connect to preprocessed data
    recon.set_input(preprocessed_data)
    # perform reconstruction
    print('reconstructing...')
    recon.process()
    # get reconstructed images
    complex_images = recon.get_output()

    # build image processing chain
    img_proc = ImagesProcessor()
    img_proc.add_gadget('g3', gadget3)
    # post-process reconstructed images
    print('processing images...')
    images = img_proc.process(complex_images)

    # show obtained images
    images.show()

try:
    main()
    print('done')

except error as err:
    # display error information
    print('??? %s' % err.value)

File: examples/demo3.m
% Lower-level interface demo, runs 3 gadget chains of different type:
% - acquisition processing chain,
% - reconstruction chain,
% - image processing chain

if ~libisloaded('mgadgetron')
    loadlibrary('mgadgetron')
end
if ~libisloaded('mutilities')
    loadlibrary('mutilities')
end

try
    % acquisitions will be read from this HDF file
    input_data = gadgetron.MR_Acquisitions('testdata.h5');
    
    % define gadgets
    gadget1 = gadgetron.Gadget('RemoveROOversamplingGadget');
    gadget2 = gadgetron.Gadget('SimpleReconGadgetSet');
    gadget3 = gadgetron.Gadget('ExtractGadget');
    
    % build acquisitions pre-processing chain
    acq_proc = gadgetron.AcquisitionsProcessor();
    acq_proc.add_gadget('g1', gadget1)
    fprintf('processing acquisitions...\n')
    interim_data = acq_proc.process(input_data);
	
    % build reconstruction chain
    recon = gadgetron.ImagesReconstructor();
	recon.add_gadget('g2', gadget2);
    % connect to input data
    recon.set_input(interim_data)
    % perform reconstruction
    fprintf('reconstructing...\n')
    recon.process()
    % get reconstructed images
    interim_images = recon.get_output();
    
    % build image post-processing chain
    proc_img = gadgetron.ImagesProcessor();
    proc_img.add_gadget('g3', gadget3);
    % post-process reconstructed images
    fprintf('processing images...\n')
    images = proc_img.process(interim_images);

    % plot obtained images
    for i = 1 : images.number()
        data = images.image_as_array(i);
        figure(1000000 + i)
        data = data/max(max(max(data)));
        imshow(data(:,:,1,1));
    end
    
catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end
